File: src/HolzShots.Legacy/Drawing/Tools/ITool.vb
Imports HolzShots.Drawing.Tools.UI
Imports HolzShots.UI.Controls

Namespace Drawing.Tools
    Friend Interface ITool(Of Out TSettings As ToolSettingsBase)
        Inherits IDisposable

        ReadOnly Property ToolType As PaintPanel.ShotEditorTool
        ReadOnly Property Cursor As Cursor
        ReadOnly Property SettingsControl As ISettingsControl(Of TSettings)

        Property BeginCoordinates As Point
        Property EndCoordinates As Point

        Sub RenderFinalImage(ByRef rawImage As Image)
        Sub RenderPreview(rawImage As Image, g As Graphics)
        Sub MouseOnlyMoved(rawImage As Image, ByRef currentCursor As Cursor, e As MouseEventArgs)
        Sub MouseClicked(rawImage As Image, e As Point, ByRef currentCursor As Cursor, trigger As Control)

        Sub LoadInitialSettings()
        Sub PersistSettings()
    End Interface
End Namespace

File: src/HolzShots.LegacyNew/Drawing/Tools/ITool.cs
using HolzShots.Drawing.Tools.UI;

namespace HolzShots.Drawing.Tools;

public interface ITool<out TSettings> : IDisposable where TSettings : ToolSettingsBase
{
    ShotEditorTool ToolType { get; }
    Cursor Cursor { get; }
    ISettingsControl<TSettings> SettingsControl { get; }

    Point BeginCoordinates { get; set; }
    Point EndCoordinates { get; set; }

    void RenderFinalImage(ref Image rawImage);
    void RenderPreview(Image rawImage, Graphics g);
    void MouseOnlyMoved(Image rawImage, ref Cursor currentCursor, MouseEventArgs e);
    void MouseClicked(Image rawImage, Point e, ref Cursor currentCursor, Control trigger);

    void LoadInitialSettings();
    void PersistSettings();
}
